server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -0,0 +1,337 @@
+        }
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {
+            if (cacheScores) {
+                scorer = ScoreCachingWrappingScorer.wrap(scorer);
+            }
+            scorer = new FilterScorable(scorer) {
+                @Override
+                public void setMinCompetitiveScore(float minScore) {
+                    // Ignore calls to setMinCompetitiveScore so that if the top docs collector
+                    // wants to skip low-scoring hits, the aggs collector still sees all hits.
+                    // this is important also for terminate_after in case used when total hits tracking is early terminated.
+                }
+            };
+            if (topDocsLeafCollector != null) {
+                topDocsLeafCollector.setScorer(scorer);
+            }
+            if (aggsLeafCollector != null) {
+                aggsLeafCollector.setScorer(scorer);
+            }
+            this.scorer = scorer;
+        }
+
+        @Override
+        public void collect(int doc) throws IOException {
+            if (shouldCollectTopDocs(doc, scorer, postFilterBits)) {
+                numCollected++;
+                if (topDocsLeafCollector != null) {
+                    try {
+                        topDocsLeafCollector.collect(doc);
+                    } catch (@SuppressWarnings("unused") CollectionTerminatedException e) {
+                        topDocsLeafCollector = null;
+                        // top docs collector does not need this segment, but the aggs collector may.
+                        if (aggsLeafCollector == null) {
+                            // TODO we keep on collecting although we have nothing to collect (there is no top docs nor aggs leaf
+                            // collector).
+                            // The reason is only to set the early terminated flag to the QueryResult. We should fix this.
+                            if (terminateAfter == 0) {
+                                throw e;
+                            }
+                        }
+                    }
+                }
+            }
+            // min_score is applied to aggs as well as top hits
+            if (isDocWithinMinScore(scorer)) {
+                if (aggsLeafCollector != null) {
+                    try {
+                        aggsLeafCollector.collect(doc);
+                    } catch (@SuppressWarnings("unused") CollectionTerminatedException e) {
+                        aggsLeafCollector = null;
+                        // aggs collector does not need this segment, but the top docs collector may.
+                        if (topDocsLeafCollector == null) {
+                            // TODO we keep on collecting although we have nothing to collect (there is no top docs nor aggs leaf
+                            // collector).
+                            // The reason is only to set the early terminated flag to the QueryResult. We should fix this.
```

Comment:
Do we have a GitHub issue for this TODO? Would this be a good FixItFriday ticket? (Maybe a good one for someone new to this part of the codebase to how it works better? Or is fixing this going to be complex refactoring?)

Reply:
This is a great idea Michael, I opened #97269. I did have to overcome the fear of adding up to our current backlog, but I see the flip side of allowing others to dive into this part of the codebase which is central to the search execution. Thanks for bringing this up.